editor/plugins/editor_context_menu_plugin.cpp

```diff
@@ -0,0 +1,210 @@
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "editor_context_menu_plugin.h"
+#include "scene/gui/popup_menu.h"
+#include "scene/resources/texture.h"
+
+bool EditorContextMenuPlugin::can_handle(const Vector<String> &p_paths) {
+	bool success = false;
+	GDVIRTUAL_CALL(_can_handle, p_paths, success);
+	return success;
+}
+
+void EditorContextMenuPlugin::add_context_menu_item(const Ref<Texture2D> &p_texture, const String &p_name, const Callable &p_callable, ContextSubMenuSlot p_submenu) {
+	Array parameters;
+	parameters.push_back(p_texture);
+	parameters.push_back(p_name);
+	parameters.push_back(p_callable);
+	set_meta("submenu", p_submenu);
+	set_meta("parameters", parameters);
+}
+
+Ref<EditorContextMenuPlugin> EditorContextMenu::scene_tree_plugins[MAX_PLUGINS];
+Ref<EditorContextMenuPlugin> EditorContextMenu::filesystem_plugins[MAX_PLUGINS];
+Ref<EditorContextMenuPlugin> EditorContextMenu::script_editor_plugins[MAX_PLUGINS];
+
+void EditorContextMenu::add_context_menu_plugin(ContextMenuSlot slot, const Ref<EditorContextMenuPlugin> &p_plugin) {
+	Ref<EditorContextMenuPlugin> *plugin_list = get_plugin_array(slot);
+	ERR_FAIL_COND(plugin_list[MAX_PLUGINS - 1] != nullptr);
+	Ref<EditorContextMenuPlugin> plugin = p_plugin;
+
+	for (int i = 0; i < MAX_PLUGINS; ++i) {
+		if (plugin_list[i] == nullptr) {
+			plugin_list[i] = p_plugin;
+			plugin->set_meta("idx", CONTEXT_ITEM_ID_BASE + i + 1);
+			break;
+		}
+	}
+}
+
+void EditorContextMenu::remove_context_menu_plugin(ContextMenuSlot slot, const Ref<EditorContextMenuPlugin> &p_plugin) {
+	Ref<EditorContextMenuPlugin> *plugin_list = get_plugin_array(slot);
+	for (int i = 0; i < MAX_PLUGINS; ++i) {
+		if (plugin_list[i] == p_plugin.ptr()) {
+			plugin_list[i].unref();
+			plugin_list[i] = nullptr;
+			break;
+		}
+	}
+}
+
+void EditorContextMenu::_bind_methods() {
+	BIND_ENUM_CONSTANT(CONTEXT_SLOT_SCENE_TREE);
+	BIND_ENUM_CONSTANT(CONTEXT_SLOT_FILESYSTEM);
+	BIND_ENUM_CONSTANT(CONTEXT_SLOT_SCRIPT_EDITOR);
+}
+
+void EditorContextMenu::handle_plugins(ContextMenuSlot p_slot, const Vector<String> &p_paths, PopupMenu *p_popup) {
+	Ref<EditorContextMenuPlugin> *plugin_list = get_plugin_array(p_slot);
+	bool add_separator = false;
+
+	for (int i = 0; i < MAX_PLUGINS; i++) {
+		Ref<EditorContextMenuPlugin> plugin = plugin_list[i];
+		if (plugin == nullptr) {
+			break;
+		}
+		int submenu = plugin->get_meta("submenu", 0);
+		if (!plugin->can_handle(p_paths) || submenu) {
```

Comment:
```suggestion
		if (submenu || !plugin->can_handle(p_paths)) {
```
Cheaper condition first